### PR TITLE
feature: add shutdown for optimizer which will release the native resources

### PR DESCRIPTION
**Pull request by i8run**
## What changes were proposed in this pull request?

This patch add an API `shutdown` for `Optimizer` to release native resources when the training end.
When should only implement them for dnn layers and optimizer. For container, it have been implemented in `Container`.

## How was this patch tested?
Manual and Jenkins.

### CODE REVIEW

**Comment by yiheng** [comment]
Please add some unit tests

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 1]
```diff
@@ -1062,6 +1063,8 @@ class DistriOptimizer[T: ClassTag] (
 
     // unpersist the model because the next time optimize is called, new `models` will be
     // created
+    shutdown()
+    CachedModels.deleteKey(modelBroadcast.uuid)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: should this be a part of shutdonw()?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 2]
```diff
@@ -963,9 +960,11 @@ class DistriOptimizer[T: ClassTag] (
 
     prepareInput()
 
-    models = DistriOptimizer.initThreadModels(model, distDataset, criterion, state,
+    val modelsAndBroadcast = DistriOptimizer.initThreadModels(model, distDataset, criterion, state,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: (models, modelBroadcast) = ... ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala [hunk 3]
```diff
@@ -250,6 +250,26 @@ trait MklDnnLayer extends AbstractModule[Activity, Activity, Float] with MklDnnM
   def parametersWithShape(): (Array[MemoryData], Array[MemoryData]) = {
     (null, null)
   }
+
+  override def release(): Unit = {
+    val tensors: ArrayBuffer[DnnTensor[Float]] = ArrayBuffer.empty
+    List(output, gradInput).filter(_ != null).foreach { t =>
+      if (t.isTensor && t.toTensor[Float].getTensorType == MklDnnType) {
```
[the remark this answers is not in the document]
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: Yes, we have the `MklDnnType`. We can't use `isInstanceOf` here, compiler will say it's type erasure.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 2 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Scala doesn't support this style. We only can init two variables like this, such as `val (a, b) = (1, 2)`. But can't assign to two `var`s, such as `var c:Int = 0; var d: Int = 0; (c, d) = (1, 2)`.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 1 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Yes. Moving to `shutdown` makes it simple.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 4]
```diff
@@ -240,4 +240,19 @@ object CachedModels {
         }
       }
     }
+
+  def deleteKey[T: ClassTag](key: String)(implicit ev: TensorNumeric[T]): Unit =
```
[the remark this answers is not in the document]
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: I'll move this whole `object` to `private[bigdl]`.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: DnnTensor[_] is type erasure?

**Pull request merged**